File: backend/app/domain/models/event.py
from pydantic import BaseModel, Field, RootModel
from typing import Dict, Any, Literal, Optional, Union, List, get_args
from datetime import datetime
import time
import uuid
from enum import Enum
from app.domain.models.plan import Plan, Step
from app.domain.models.file import FileInfo
import json
from app.domain.models.search import SearchResultItem
from app.domain.models.image import ImageSearchResultItem
from app.domain.models.validation import ValidationResult


class PlanStatus(str, Enum):
    """Plan status enum"""
    CREATED = "created"
    UPDATED = "updated"
    COMPLETED = "completed"


class StepStatus(str, Enum):
    """Step status enum"""
    STARTED = "started"
    FAILED = "failed"
    COMPLETED = "completed"


class ToolStatus(str, Enum):
    """Tool status enum"""
    CALLING = "calling"
    CALLED = "called"


class BaseEvent(BaseModel):
    """Base class for agent events"""
    type: Literal[""] = ""
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    timestamp: datetime = Field(default_factory=lambda: datetime.now())

class ErrorEvent(BaseEvent):
    """Error event"""
    type: Literal["error"] = "error"
    error: str

class PlanEvent(BaseEvent):
    """Plan related events"""
    type: Literal["plan"] = "plan"
    plan: Plan
    status: PlanStatus
    step: Optional[Step] = None

class BrowserToolContent(BaseModel):
    """Browser tool content"""
    screenshot: str = ""
    js_code: Optional[str] = None
    js_result: Optional[Any] = None

class SearchToolContent(BaseModel):
    """Search tool content"""
    results: List[SearchResultItem]

class ShellToolContent(BaseModel):
    """Shell tool content"""
    console: Any
    # Resolved shell session id. The agent may omit the `id` argument on
    # shell_exec (recommended: the toolkit auto-creates a fresh session), so
    # the frontend needs the ACTUAL id — echoed here from the tool result —
    # to poll live console output via /sessions/{id}/shell.
    session_id: Optional[str] = None

class FileToolContent(BaseModel):
    """File tool content"""
    content: str
    # Pre-edit snapshot so the UI can show Diff / Original / Modified tabs
    # (official Manus text_editor behaviour). Only present when the file
    # already existed before file_write / file_str_replace modified it.
    old_content: Optional[str] = None

class McpToolContent(BaseModel):
    """MCP tool content"""
    result: Any

class ImageToolContent(BaseModel):
    """Image tool content — search results, download confirmation, or generated image"""
    results: List[ImageSearchResultItem] = Field(default_factory=list)
    downloaded_file: Optional[str] = None
    downloaded_file_id: Optional[str] = None
    downloaded_signed_url: Optional[str] = None
    generated_url: Optional[str] = None
    generated_prompt: Optional[str] = None
    generated_model: Optional[str] = None

ToolContent = Union[
    BrowserToolContent,
    SearchToolContent,
    ShellToolContent,
    FileToolContent,
    McpToolContent,
    ImageToolContent,
]

class ToolEvent(BaseEvent):
    """Tool related events"""
    type: Literal["tool"] = "tool"
    tool_call_id: str
    tool_name: str
    tool_content: Optional[ToolContent] = None
    function_name: str
    function_args: Dict[str, Any]
    status: ToolStatus
    function_result: Optional[Any] = None
    # Official Manus StandardToolUsed prefers ``brief`` (natural-language
    # action label supplied by the model) over raw paths / commands.
    brief: Optional[str] = None

class TitleEvent(BaseEvent):
    """Title event"""
    type: Literal["title"] = "title"
    title: str

class StepEvent(BaseEvent):
    """Step related events"""
    type: Literal["step"] = "step"
    step: Step
    status: StepStatus

class MessageEvent(BaseEvent):
    """Message event"""
    type: Literal["message"] = "message"
    role: Literal["user", "assistant"] = "assistant"
    message: str
    attachments: Optional[List[FileInfo]] = None
    # True ONLY for the task's final summary message. Files created during
    # the task are never attached to mid-task messages — they are deferred
    # and delivered exactly once, here, with the final summary (Manus-style).
    is_final: bool = False
    # True for compact progress narrations (tool-progress lines, step
    # transitions). The frontend renders them INSIDE the unified step
    # timeline instead of standalone chat bubbles.
    is_progress: bool = False
    # True ONLY for agent questions that require a user answer (ask_user).
    # These must stay VISIBLE as standalone chat messages — they pause the
    # task — so the timeline must not swallow them.
    is_question: bool = False

class DoneEvent(BaseEvent):
    """Done event"""
    type: Literal["done"] = "done"

class WaitEvent(BaseEvent):
    """Wait event"""
    type: Literal["wait"] = "wait"

class MessageChunkEvent(BaseEvent):
    """Streaming message chunk — emitted token-by-token for the acknowledgment.
    The frontend accumulates these into a single message bubble.
    When done=True the stream for this message is complete."""
    type: Literal["message_chunk"] = "message_chunk"
    role: Literal["user", "assistant"] = "assistant"
    content: str = ""
    done: bool = False

class ValidationEvent(BaseEvent):
    """Final validation gate result — emitted once, right before the task's
    final summary message. Carries the mechanical checks, the execution
    summary counts, and the evidence register collected from the run."""
    type: Literal["validation"] = "validation"
    result: ValidationResult

AgentEvent = Union[
    ErrorEvent,
    PlanEvent, 
    ToolEvent,
    StepEvent,
    MessageEvent,
    MessageChunkEvent,
    ValidationEvent,
    DoneEvent,
    TitleEvent,
    WaitEvent,
]

File: frontend/src/types/event.ts
import type { FileInfo } from '../api/file';

export type AgentSSEEvent = {
  event: 'tool' | 'step' | 'message' | 'message_chunk' | 'error' | 'done' | 'title' | 'wait' | 'plan' | 'attachments' | 'validation' | 'knowledge';
  data: ToolEventData | StepEventData | MessageEventData | MessageChunkEventData | ErrorEventData | DoneEventData | TitleEventData | WaitEventData | PlanEventData | ValidationEventData | KnowledgeEventData;
}

export interface BaseEventData {
  event_id: string;
  timestamp: number;
}

export interface ToolEventData extends BaseEventData {
  tool_call_id: string;
  name: string;
  status: "calling" | "called";
  function: string;
  args: {[key: string]: any};
  content?: any;
  /** Official Manus natural-language action label. */
  brief?: string;
}

export interface StepEventData extends BaseEventData {
  status: "pending" | "running" | "completed" | "failed"
  id: string
  description: string
  /** Present when the step finished with a concrete outcome. */
  result?: string
}

export interface MessageEventData extends BaseEventData {
  content: string;
  role: "user" | "assistant";
  attachments: FileInfo[];
  /** true = compact progress narration belonging to the step timeline */
  is_progress?: boolean;
  /** true = the task's final summary message (single delivery point for files) */
  is_final?: boolean;
  /** true = agent question that pauses the task and needs a user answer */
  is_question?: boolean;
}

export interface MessageChunkEventData extends BaseEventData {
  content: string;
  role: "user" | "assistant";
  done: boolean;
}

export interface ErrorEventData extends BaseEventData {
  error: string;
}

export interface DoneEventData extends BaseEventData {
}

export interface WaitEventData extends BaseEventData {
}

export interface TitleEventData extends BaseEventData {
  title: string;
}

export interface PlanEventData extends BaseEventData {
  steps: StepEventData[];
}

// ── Final validation gate (P0) ────────────────────────────────────────────
export type CheckState = 'pass' | 'fail' | 'warn' | 'skipped';

export interface ValidationCheck {
  key: string;
  state: CheckState;
  detail: string;
}

export interface EvidenceEntry {
  id: string;
  summary: string;
  url: string;
  requested_url?: string;
  title: string;
  site_name?: string;
  published_date?: string | null;
  quote?: string;
  verified: boolean;
  source: 'search' | 'browser';
  redirected?: boolean;
}

export interface ExecutionSummaryData {
  started_at?: string | null;
  finished_at?: string | null;
  duration_seconds?: number | null;
  total_steps: number;
  steps_completed: number;
  steps_failed: number;
  tool_calls_total: number;
  tool_calls_succeeded: number;
  tool_calls_failed: number;
  files_created: number;
  files_updated: number;
  evidence_count: number;
  warnings: number;
  errors: number;
}

export interface ValidationResultData {
  overall: 'pass' | 'needs_review';
  checks: ValidationCheck[];
  unresolved_errors: number;
  warnings: number;
  summary: ExecutionSummaryData;
  evidence: EvidenceEntry[];
}

export interface ValidationEventData extends BaseEventData {
  result: ValidationResultData;
}
export interface KnowledgeEventData extends BaseEventData {
  /** Proposed learnings (display text). */
  items: string[];
  /** Knowledge item ids aligned with items by index (accept/reject API). */
  item_ids?: string[];
  status: 'pending';
}
